src/doc/trpl/closures.md

```diff
@@ -1,214 +1,510 @@
-fn  add_one   (x: i32) -> i32 { 1 + x }
+assert_eq!(2, plus_one(1));
 ```
 
-As you may have noticed, closures infer their argument and return types, so you
-don't need to declare one. This is different from named functions, which
-default to returning unit (`()`).
+But we don't have to. Why is this? Basically, it was chosen for ergonomic reasons.
+While specifying the full type for named functions is helpful, closures aren't going
+to, for example, appear in Rustdoc output somewhere. They don't cause the kinds of
+error-at-a-distance that inferring named function types can. In practice, it
+works well.
 
-There's one big difference between a closure and named functions, and it's in
-the name: a closure "closes over its environment." What does that mean? It means
-this:
+The second is that the syntax is similar, but a bit different. I've added spaces
+here to make them look a little closer:
 
-```{rust}
+```rust
+fn  plus_1 (  x: i32 ) -> i32 { x + 1 }
+let plus_2 = |x: i32 | -> i32 { x + 1 };
+let plus_3 = |x      |          x + 1  ;
+```
+
+Small differences, but they're similar in ways.
+
+# Closures and their environment
+
+As we said, closures are called such because they 'close over their environment.' It
+looks like this:
+
+```rust
+let num = 5;
+let plus_num = |x| x + num;
+
+assert_eq!(10, plus_num(5));
+```
+
+This closure, `plus_num`, refers to a `let` binding in its scope: `num`. More
+specifically, it borrows the binding. If we do something that would conflict
+with that binding, we get an error. Like this one:
+
+```rust,ignore
+let mut num = 5;
+let plus_num = |x| x + num;
+
+let y = &mut num;
+```
+
+Which errors with:
+
+```text
+error: cannot borrow `num` as mutable because it is also borrowed as immutable
+    let y = &mut num;
+                 ^~~
+note: previous borrow of `num` occurs here due to use in closure; the immutable
+  borrow prevents subsequent moves or mutable borrows of `num` until the borrow
+  ends
+    let plus_num = |x| x + num;
+                   ^~~~~~~~~~~
+note: previous borrow ends here
 fn main() {
-    let x: i32 = 5;
+    let mut num = 5;
+    let plus_num = |x| x + num;
+    
+    let y = &mut num;
+}
+^
+```
+
+A verbose yet helpful error message! As it says, we can't take a mutable borrow
+on `num` because the closure is already borrowing it. If we let the closure go
+out of scope, we can:
+
+```rust
+let mut num = 5;
+{
+    let plus_num = |x| x + num;
+
+} // plus_num goes out of scope, borrow of num ends
+
+let y = &mut num;
+```
+
+If your closure requires it, however, Rust will take ownership and move
+the environment instead:
 
-    let printer = || { println!("x is: {}", x); };
+```rust,ignore
+let nums = vec![1, 2, 3];
 
-    printer(); // prints "x is: 5"
+let takes_nums = || nums;
+
+println!("{:?}", nums);
+```
+
+This gives us:
+
+```text
+note: `nums` moved into closure environment here because it has type
+  `[closure(()) -> collections::vec::Vec<i32>]`, which is non-copyable
```

Comment:
if this is really the error message, it seems buggy. it should be saying `Vec<i32>` or something I expect?